vochain/state.go

```diff
@@ -504,39 +520,103 @@ func (v *State) IsOracle(addr common.Address) (bool, error) {
 }
 
 // setTreasurer saves the Treasurer address to the state
-func (v *State) setTreasurer(address common.Address) error {
+func (v *State) SetTreasurer(address common.Address) error {
+	tBytes, err := proto.Marshal(
+		&models.Treasurer{
+			Address: address.Bytes(),
+			Nonce:   0,
+		},
+	)
+	if err != nil {
+		return err
+	}
 	v.Tx.Lock()
 	defer v.Tx.Unlock()
-	return v.Tx.DeepSet([]byte(TreasurerKey), address.Bytes(), ExtraCfg)
+	return v.Tx.DeepSet([]byte(TreasurerKey), tBytes, ExtraCfg)
 }
 
-func (v *State) Treasurer(isQuery bool) (common.Address, error) {
+// Treasurer returns the address and the Treasurer nonce
+func (v *State) Treasurer(isQuery bool) (*models.Treasurer, error) {
 	if !isQuery {
 		v.Tx.RLock()
 		defer v.Tx.RUnlock()
 	}
-
 	extraTree, err := v.mainTreeViewer(isQuery).SubTree(ExtraCfg)
 	if err != nil {
-		return common.Address{}, err
+		return nil, err
 	}
-
 	var rawTreasurer []byte
 	if rawTreasurer, err = extraTree.Get([]byte(TreasurerKey)); err != nil {
-		return common.Address{}, err
+		return nil, err
+	}
+	var t models.Treasurer
+	if err := proto.Unmarshal(rawTreasurer, &t); err != nil {
+		return nil, err
 	}
-	return common.BytesToAddress(rawTreasurer), nil
+	return &t, nil
 }
 
+// IsTreasurer returns true if the given address matches the Treasurer address
 func (v *State) IsTreasurer(addr common.Address) (bool, error) {
-	tAddr, err := v.Treasurer(false)
+	t, err := v.Treasurer(false)
 	if err != nil {
 		return false, err
 	}
-
+	tAddr := common.BytesToAddress(t.Address)
 	return addr == tAddr, nil
 }
 
+// incrementTreasurerNonce increments the treasurer nonce
+func (v *State) incrementTreasurerNonce() error {
+	t, err := v.Treasurer(false)
+	if err != nil {
+		return err
```

Comment:
Ideally all these errors should be wrapped. Else its very hard to identify where it comes from